src/Common/src/Interop/Unix/System.Native/Interop.readdir.cs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class Sys
+    {
+        private static readonly int s_direntSize = GetDirentSize();
+
+        internal enum NodeType : short
+        {
+            DT_UNKNOWN  =  0,
+            DT_FIFO     =  1,
+            DT_CHR      =  2,
+            DT_DIR      =  4,
+            DT_BLK      =  6,
+            DT_REG      =  8,
+            DT_LNK      = 10,
+            DT_SOCK     = 12,
+            DT_WHT      = 14
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private unsafe struct InternalDirectoryEntry
+        {
+            internal NodeType   InodeType;
+            internal IntPtr     Name;
+            internal int        NameLength;
+        }
+
+        internal struct DirectoryEntry
+        {
+            internal NodeType   InodeType;
+            internal string     InodeName;
+        }
+
+        [DllImport(Libraries.SystemNative, SetLastError = true)]
+        internal static extern Microsoft.Win32.SafeHandles.SafeDirectoryHandle OpenDir(string path);
+
+        [DllImport(Libraries.SystemNative, SetLastError = false)]
+        internal static extern int GetDirentSize();
+
+        [DllImport(Libraries.SystemNative, SetLastError = true)]
+        private static unsafe extern int ReadDirR(SafeDirectoryHandle dir, byte* buffer, int bufferSize, out InternalDirectoryEntry outputEntry);
+
+        [DllImport(Libraries.SystemNative, SetLastError = true)]
+        internal static extern int CloseDir(IntPtr directory);
+
+        // The calling pattern for ReadDir is described in src/Native/System.Native/pal_readdir.cpp
+        internal static int ReadDir(SafeDirectoryHandle dir, out DirectoryEntry outputEntry)
+        {
+            unsafe
+            {
+                // To reduce strcpys, alloc a buffer here and get the result from OS, then copy it over for the caller.
+                byte* buffer = stackalloc byte[s_direntSize];
+                InternalDirectoryEntry temp = new InternalDirectoryEntry();
+                int ret = ReadDirR(dir, buffer, s_direntSize, out temp);
+                if (ret == 0)
+                {
+                    outputEntry = new DirectoryEntry() { InodeName = GetDirectoryEntryName(temp), InodeType = temp.InodeType };
+                }
+                else
+                {
+                    outputEntry = default(DirectoryEntry);
+                }
```

Comment:
Just a style thing, and you can leave it as-is if you prefer, but I typically like to write things like this like:
```C#
outputEntry = ret == 0 ?
    new DirectoryEntry { InodeName = GetDirectoryEntryName(temp), InodeType = temp.InodeType } :
    default(DirectoryEntry);
```
in part to be more concise, but more importantly to highlight that we're always initializing ```outputEntry```, and how we do so is just conditional on the value of  ```ret```.